src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,311 @@
+    """
+    mu = mu if isinstance(mu, Mu) else parameter_space.parameters.parse(mu)
+    active = np.array([])
+
+    for key, dim in parameter_space.parameters.items():
+        indices = np.zeros(dim)
+        low_range = parameter_space.ranges[key][0] * np.ones(dim)
+        high_range = parameter_space.ranges[key][1] * np.ones(dim)
+        eps = epsilon * np.ones(dim)
+        indices[np.where(np.logical_or(mu[key] - low_range <= eps, high_range - mu[key] <= eps))] = 1
+        active = np.concatenate((active, indices))
+
+    inactive = np.ones(active.shape) - active
+
+    return active, inactive
+
+
+def compute_hessian_action_matrix(hessian, direction, active, inactive):
+    """Project the Hessian matrix applied to the direction using an active and an inactive set.
+
+    Parameters
+    ----------
+    hessian
+        The Hessian matrix.
+    direction
+        The descent direction used to update the parameter.
+    active
+        The active set computed from the |ParameterSpace|.
+    inactive
+        The inactive set computed from the |ParameterSpace|.
+
+    Returns
+    -------
+    hessian
+        The projection of the Hessian applied to the direction.
+    """
+    direction_A = np.multiply(active, direction)
+    direction_I = np.multiply(inactive, direction)
+    hessian = direction_A + np.multiply(inactive, hessian.dot(direction_I))
+    return hessian
+
+
+def update_hessian(hessian, mu, old_mu, gradient, old_gradient):
+    """Update Hessian matrix using BFGS iteration.
+
+    Parameters
+    ----------
+    hessian
+        The current Hessian matrix.
+    mu
+        The current `mu` parameter.
+    old_mu
+        The previous `mu` parameter.
+    gradient
+        The current gradient with respect to the parameter.
+    old_gradient
+        The previous gradient with respect to the parameter.
+
+    Returns
+    -------
+    hessian
+        The next BFGS Hessian matrix iterate.
+    """
+    gradient_difference = gradient - old_gradient
+    mu_difference = mu - old_mu
+    gradient_mu_coefficient = np.dot(gradient_difference, mu_difference)
+
+    if gradient_mu_coefficient > 0.:
+        hessian_grad_diff = np.dot(hessian, gradient_difference)
+        quadratic_coefficient = np.dot(gradient_difference, hessian_grad_diff)
+
+        hessian_part = np.outer(hessian_grad_diff, mu_difference) + np.outer(mu_difference, hessian_grad_diff)
+        mu_part = np.outer(mu_difference, mu_difference)
+
+        mu_coeff = (gradient_mu_coefficient + quadratic_coefficient) / gradient_mu_coefficient**2
+        hessian_coeff = 1. / gradient_mu_coefficient
+
+        hessian += mu_coeff * mu_part - hessian_coeff * hessian_part
+    else:
+        hessian = np.eye(hessian.shape[0])
+
+    return hessian
+
+
+@defaults('miniter', 'maxiter', 'rtol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def bfgs(model, parameter_space, initial_guess=None, miniter=0, maxiter=100, rtol=1e-16,
```

Comment:
Without the `error_aware`-part, I completely agree that it would not be necessary to have it.
The main thing why this BFGS is different is the (very unusual) additional line search criterion which is called `error_criterion` below in the code. Back in the days, I did not manage to incorporate this into the `scipy` code although it is basically only a function call.

However, it might be that the criterion nowadays can be incorporated to the BFGS within the `constraints` argument in `scipy.optimize.minimize`. @peoe, have you looked into this? Might be worth a check. Or do you see another reason why we write our own BFGS?

Reply:
As far as I can tell, SciPy only utilizes the constraints parameter for ["Only for COBYLA, SLSQP and trust-constr."](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize). So unless we were to perform some intrusive things during runtime or always manually treat the data after the algorithm to fit into our framework I don't see how we might incorporate the additional criterion with scipy...

Besides avoiding potential dependency troubles further down the line, another minor argument is that if one were to extend the TR/BFGS code (e. g. to the non-conforming dual setting that Tim covered or for parabolic problems, ...), then providing a standalone implementation will be much simpler to work with.

Reply:
I just had another look at SciPy. It definitely does not work with what is implemented there. The inequality constraint that we would need can only be treated within the 'SLSQP' which heavily iterates around the boundary of the trust-region (which we do not allow in our algorithm).

Of course contributing to SciPy is an option, but I honestly do not know any application where it makes sense to terminate the sub-problem earlier instead of iterating across the boundary. Especially because we are not interested in implementing the TR algorithm but only the error-aware sub-problem. The feature request would go down to changing the [armijo line search](https://github.com/scipy/scipy/blob/9aee554b9f87d53eb32fd2b0318c45439c9d9950/scipy/optimize/_linesearch.py#L669) in SciPy which I do not think they would be much interested in.

For reference: I used the following lines of code for the subproblem in `pymor/algorithms/tr.py`

```python 
with logger.block(f'Solving subproblem for mu {mu} with BFGS...'):
    # BFGS with error_aware criterion as we use it now ! 
    mu, sub_data = bfgs(
        surrogate.rom, parameter_space, initial_guess=mu, miniter=miniter_subproblem,
        maxiter=maxiter_subproblem, rtol=rtol, tol_sub=tol_sub,
        line_search_params=line_search_params, stagnation_window=stagnation_window,
        stagnation_threshold=stagnation_threshold, error_aware=True,
        error_criterion=error_aware_line_search_criterion, beta=beta, radius=radius)

    # BFGS without error_aware criterion
    mu_, sub_data_ = bfgs(
        surrogate.rom, parameter_space, initial_guess=old_mu, miniter=miniter_subproblem,
        maxiter=maxiter_subproblem, rtol=rtol, tol_sub=tol_sub,
        line_search_params=line_search_params, stagnation_window=stagnation_window,
        stagnation_threshold=stagnation_threshold, error_aware=False,
        error_criterion=error_aware_line_search_criterion, beta=beta, radius=radius)
    
    # BFGS in scipy without error_aware criterion
    def objective_functional(mu_in):
        return surrogate.rom.output(mu_in)[0, 0]

    def gradient_of_functional(mu_in):
        return surrogate.rom.output_d_mu(surrogate.rom.parameters.parse(mu_in), return_array=True)
    
    result = minimize(objective_functional,
                      old_mu,
                      method='L-BFGS-B',
                      jac=gradient_of_functional,
                      bounds=(parameter_space.ranges['diffusion'], parameter_space.ranges['diffusion']),
                      options={'ftol': rtol, 'gtol': tol_sub})

    def error_aware_line_search_criterion_for_scipy(new_mu):
        current_value = surrogate.rom.output(new_mu)[0, 0]
        output_error = surrogate.estimate_output_error(new_mu)
        return output_error / abs(current_value)
    
    # SLSQP from scipy with error_aware criterion
    from scipy.optimize import NonlinearConstraint
    const = NonlinearConstraint(error_aware_line_search_criterion_for_scipy, 0, beta * radius)

    result_ = minimize(objective_functional,
                      old_mu,
                      method='SLSQP',
                      jac=gradient_of_functional,
                      bounds=(parameter_space.ranges['diffusion'], parameter_space.ranges['diffusion']),
                      options={'ftol': rtol, 'gtol': tol_sub},
                      constraints=const)
```
Looking at the results in this specific example:
Our self-written BFGS takes 2 iterations before the boundary is reached. The SLSQP instead takes 36 iterations until it finishes.
This already shows that iterating across the boundary is not feasible for our algorithm.

Maybe as a side note: In order to make clear that the BFGS is special in that regard, I would call it `error_aware_BFGS` or something similar.

Reply:
Ok, then I think it would be good to add an explanation in the `bfgs` docstring about what makes it special. Renaming the function also sounds good.

Reply:
I must say that I do not completely understand what is available in scipy compared to what is implemented here. Regarding BFGS, there also is a `BFGS` `HessianUpdateStrategy` which one might consider to use. Then, there are also implementations of trust region methods in scipy. In particular, there is also the (somewhat private) `_minimize_trust_region` method, which you can pass a `subproblem`. Have you looked into that?

Reply:
Thanks @sdrave  for pointing into this direction.

AFAIK, `BFGS-B` (`B` for bounds) is only implemented within the `L-BFGS-B` (`L` for limited-memory) variant. The `L-BFGS-B` is Fortran code which I do not intend to look into. Also: I do not see the point of using it, since these are only 3 relatively simple methods in our case.

Regarding the `_minimize_trust_region` methods in `scipy`, I want to remind you that our method is mathematically very different to what is done in `_minimize_trust_region` and can not be used. Moreover, a generalization to our liking would probably be of low interest to `scipy`. Apart from this, we would have to implement our own `subproblem` (which then would have code that is currently in our `BFGS` method).

Just to make this point clear: The Trust-Region methods in scipy all have in common that the `trust-region` is determined by a metric distance, not an actual `error_estimator`. Moreover, the cheap model function for the subproblem is constructed from scratch, whereas, in our method, the model function is reused but updated.

Reply:
Ok. Although I think that it's still kind of a shame that we have to reimplement everything, I see that it is probably not feasible to get the scipy code adapted to our needs.

> Just to make this point clear: The Trust-Region methods in scipy all have in common that the trust-region is determined by a metric distance, not an error_estimator. Moreover, the cheap model function for the subproblem is constructed from scratch for each outer iteration, whereas, in our method, the model function is reused but updated.

Some summary like that should definitely go into the docstring.

Reply:
> Ok. Although I think that it's still kind of a shame that we have to reimplement everything, I see that it is probably not feasible to get the scipy code adapted to our needs.

I do not understand the "shame". It is a completely different method that is no use for `scipy` unless they want to do model reduction, which is our thing right?

> 
> > Just to make this point clear: The Trust-Region methods in scipy all have in common that the trust-region is determined by a metric distance, not an error_estimator. Moreover, the cheap model function for the subproblem is constructed from scratch for each outer iteration, whereas, in our method, the model function is reused but updated.
> 
> Some summary like that should definitely go into the docstring.

I agree !

Reply:
> I do not understand the "shame". It is a completely different method that is no use for `scipy` unless they want to do model reduction, which is our thing right?

I wouldn't agree that the method is 'completely' different, and I would assume that you can think of other surrogates, which do not come from MOR as we understand ist. However, let's have that discussion somewhere else.

I'll try to do a full review today or tomorrow.

Reply:
To summarize the changes I've just added (I hope I didn't skip over any feedback):

Reply:
To summarize the changes I've just added (I hope I didn't miss any of your feedback):
* renamed `bfgs` to `error_aware_bfgs`
* added a remark to the BFGS docstring about terminating close to the TR boundary, contrasting the `scipy.optimize` behaviour
* added a remark to the TR docstring about computation of TR conditions and construction of the model surrogate